### PR TITLE
Enhance queryClient with pycurl streaming support

### PR DESCRIPTION
**Pull request by iglesu**
- Implement pycurl-based data streaming in queryClient if pycurl is installed, optimizing data retrieval from the backend.
- Extend datalab shell interface to support data streaming to files.
- Update requirements.txt for Python 3.10 and 3.11 compatibility, and introduce symlinks for requirements files (requirements_<python_version>.txt) to indicate supported Python versions.
- Modify setup.py to inform users about the Python versions supported by datalab, guided by the available requirements file symlinks.
- Provide a Docker testing and usage option, offering an alternative setup method for users.

Related changes also in querymanager:
https://github.com/astro-datalab/querymanager/commit/5ceff04c

### CODE REVIEW

File: README.md
Comment (chaduhduh) [inline README.md:246]: Is docker required to run the tests? We could maybe document what the testing would look like without docker (although I will be one of the Docker users :smile_cat: ). I think this suggestion wouldn't be related to what you added though .. it is just documentation we should have had already I suppose. Maybe a new issue for that?

File: requirements_3.10.txt
Comment (chaduhduh) [inline requirements_3.10.txt:1]: I don't think I fully understand what these files are for. We may want to consider if they can be replaced by an update to the "System Requirements" section of the README. If we leave them we should maybe add documentation about the purpose and what to do with them. Is there some implicit behavior going on with these that I am unaware of perhaps?

File: tests/docker_test.sh
Comment (chaduhduh) [inline tests/docker_test.sh:48]: This might loop forever if the phase is `ERROR` or `ABORTED`

File: tests/docker_test.sh
Comment (chaduhduh) [inline tests/docker_test.sh:68]: This could exit with some error code in addition to the error message output.

**Review by chaduhduh — CHANGES_REQUESTED** [review]
Hey @iglesu . This looks fine and I think a lot of the changes were already reviewed in the referenced PR.

I marked as requesting changes mainly just for the leftover comments. I think everything else can be incporated at your discretion and feel free to comment on the threads as you see fit.

File: requirements_3.10.txt
Comment (rnikutta) [reply requirements_3.10.txt:1]: I, too, don't understand the need for Python version-specific requirements.txt files. Also, this is likely rather non-standard.

File: requirements.txt
Comment (rnikutta) [inline requirements.txt:1]: The previous versions were *very* restrictive. The new ones are broader, but still very narrow. In some cases that certainly needed, but can't we relax some of them a bit more. For instance scipy, numpy, astropy? (astropy v6 is out since a while...) and others?

File: README.md
Comment (rnikutta) [reply README.md:246]: Mark the entire section "Docker Tests" clearly as "totally optional". Users reading the README file should be absolutely clear that that don't have to consider the section at all if they don't want to. (always think of users as interested in running a tool to get data, not to do testing of any kind).

File: tests/docker_test.sh
Comment (rnikutta) [reply tests/docker_test.sh:48]: i.e. needs a global timeout. Should be the same as the backend timeout probably (whose value I don't recall; 6 hours? 24 hours? @mjfitzpatrick ?

Comment (chaduhduh) [reply tests/docker_test.sh:48]: I don't think a global timeout is needed for this unless I am missing something? My comment above was about refactoring this test script a bit.

File: requirements_3.10.txt
Comment (iglesu) [reply requirements_3.10.txt:1]: Ok, I'll ditch that, I'll hard code the python version we support.

File: README.md
Comment (iglesu) [reply README.md:246]: The docker thing was more of a tool towards a possible CI workflow but also, or so I thought, maybe something useful to the users? If it creates more confusion than help I'll take it away.

File: tests/docker_test.sh
Comment (rnikutta) [inline tests/docker_test.sh:62]: This looks like it could be slower than just ``line_count=$((`time wc -l < $csv_file`-1))``
(some local testing seems to suggest that)

File: README.md
Comment (rnikutta) [reply README.md:246]: Or consider a separate .md file for this, maybe TESTING.md (or something that's quasi-standard out there).

File: tests/docker_test.sh
Comment (iglesu) [reply tests/docker_test.sh]: good catch!

Comment (rnikutta) [reply tests/docker_test.sh]: Sorry, my comment was regarding the `while [[ "$STATUS" != "COMPLETED" ]]` ... If for some reason the status never reaches COMPLETED, this should still time out some time.

File: README.md
Comment (chaduhduh) [reply README.md:246]: If we are going to have test files and docker files in this repository then we should document them in some form. I think creating another markdown file as a "Development Guide" would make sense so we can separate some of these usage and development details. This would also address Robert's concern. If we go this route we then we should include something like in my original comment (as new copy in that file or as a new issue :ok_hand: )

Whether or not we should have these files in here is another question. My thought is that it doesn't seem to hurt anything having these docker and tests files there and might make it easier for others to contribute and for us to develop workflows (as you pointed out).

Comment (rnikutta) [reply README.md:246]: Agree with @chaduhduh but I feel we need to think about it a bit more first. @iglesu do you mind removing this from this branch/PR, and open an issue, e.g. labeled "Add developer info", where we flesh this out (which would be public b/c of the repo being public), or maybe better, a Jira issue?

File: tests/docker_test.sh
Comment (chaduhduh) [reply tests/docker_test.sh]: > Sorry, my comment was regarding the `while [[ "$STATUS" != "COMPLETED" ]]` ... If for some reason the status never reaches COMPLETED, this should still time out some time.

Ahh ok! I I'll defer to @iglesu to consider if also having a timeout or max iterations is appropriate. I think we can reliably check for all of the phases that represent that the job is done (with or without error) which is what my comment was about but theoretically it could also get stuck in a `QUEUED` or `EXECUTING` phase which I think is what Robert was getting at. Maybe not an edge case we need to handle in this context but also it probably wouldn't be terribly difficult to implement into the existing logic.

Comment (iglesu) [reply tests/docker_test.sh]: > Sorry, my comment was regarding the `while [[ "$STATUS" != "COMPLETED" ]]` ... If for some reason the status never reaches COMPLETED, this should still time out some time.

This issue has been fixed in this commit: 64e50868bccd6a0d9a620497f5af7c3b3a9f8563

File: tests/docker_test.sh
Comment (iglesu) [reply tests/docker_test.sh:68]: also fix in commit 64e50868bccd6a0d9a620497f5af7c3b3a9f8563

File: tests/docker_test.sh
Comment (iglesu) [reply tests/docker_test.sh]: but it is more beautiful! and if beauty costs 0.0000000001% extra of CPU, so be it! LoL!

**Comment by iglesu** [comment]
@chaduhduh @rnikutta I removed the docker/test stuff,  as I see it generated a PR a lot more active than I expected. LoL! Anyway, I think with the latest commit, the state of the PR maintains the status quo. And the basic point I get, and I agree with, is that the status quo should stay for this bug fix, which is what it is, just a bug fix. As for any further changes, I think it should be Robert who makes the executive decisions on our next steps. Given our time constraints and the FTE-s and work-load we're handling, we really can't afford to devote time to non-priorities, especially not to Byzantine discussions. Apologies for the lengthy PR.

**Review by chaduhduh — APPROVED** [review]
Looks good :confetti_ball:

File: tests/docker_test.sh
Comment (rnikutta) [reply tests/docker_test.sh]: Please test yourself, but I get a factor ~3 in speed difference on an ~800 MB CSV file...

**Comment by iglesu** [comment]
noted, thank you Robert!

**Comment by iglesu** [comment]
I added @jacquesalice PR #68 changes to this PR as there were quite a bit of entangled code and it was easier to move everything to this PR.

In what refers to the pycurl download functionality I found the `certifi` package which is precisely provides its own safe CA certificate, which will solve the installation problems.

So if you could install and test it in your machines it'll be such a big help, please follow installation documentation and please report back anything that's not clear.

I've added below some example code to make easier for you to test:

```
import getpass                          # for Data Lab login
import dl
from dl import authClient               # for Data Lab login
from dl import queryClient as qc        # Query Manager client interface
import time

token = authClient.login(input("Please, enter username: "), getpass.getpass('Account password: '))
print(token)

if not authClient.isValidToken(token):
    print('Error: invalid user login (%s)' % token)
    exit(1)

print(f"query svc url:{qc.get_svc_url()}")

try:
    
    query = 'SELECT id,ndet FROM nsc_dr2.object WHERE ndet>=50 LIMIT 100'
    job_id = qc.query(query, async_=True)
    print(f"jobid:{job_id}")

except Exception as e:
    print("Exception:", str(e))

status = qc.status(job_id)
print(f"async job status:{status}")

# Loop until status is either ERROR or COMPLETED
while status not in ['ERROR', 'COMPLETED']:
    time.sleep(10)  # Wait for a bit before checking again to avoid spamming requests
    status = qc.status(job_id)
    print(f"async job status:{status}")

#try to get results
print(f"getting results for {job_id}")
st_results = qc.results(token, job_id)

print(st_results)
```




 @jacquesalice @chaduhduh @rnikutta @mjfitzpatrick  Thank you!!

**Comment by jacquesalice** [comment]
I'm running into an error when importing queryClient and pycurl. Here's what I did:
1. Created python 3.10 environment:
```
conda create -n pyenv3.10_queryclient python=3.10
conda activate pyenv3.10_queryclient
```
2. Cloned the `datalab` repo, switched to the `query_async_download` branch
3. Ran the following from the README:
```
python -m pip install --upgrade pip setuptools
python -m pip install build
python -m build
pip install --user dist/astro_datalab-2.23.0-py3-none-any.whl
touch ~/.netrc
```
4. Created a file called `asynctest.py` with the contents from Igor's example code comment above
5. Tried to run `asynctest.py`, but got an error:
```
(pyenv3.10_queryclient) alice.jacques@Alice-Jacques-MacBook-Pro-2 datalab % ./asynctest.py 
Traceback (most recent call last):
  File "/Users/alice.jacques/datalab_repo/datalab/./asynctest.py", line 5, in <module>
    from dl import queryClient as qc        # Query Manager client interface
  File "/Users/alice.jacques/datalab_repo/datalab/dl/queryClient.py", line 115, in <module>
    import pycurl
ImportError: pycurl: libcurl link-time version (7.87.0) is older than compile-time version (8.4.0)
```

**Comment by rnikutta** [comment]
@iglesu Congratulations on this one! I installed from sources and ran a few tests, from 100 rows x 2 columns, all the way up to 10M rows x 10 columns (1.1GB CSV). It downloads the results like a champ each time. Even for the largest files it was averaging over 6 MBytes/second all the way. Wow.

**Review by rnikutta — APPROVED** [review]
Sorry, forgot to approve!

**Comment by jacquesalice** [comment]
Ok I upgraded my Mac OS from Monterey 12.6.6 to Sonoma 14.5 and now it works! The docstrings for `queryClient.py` also look good here, so I'll close PR #68 . Thanks @iglesu !

**Review by jacquesalice — APPROVED** [review]
Ok I upgraded my Mac OS from Monterey 12.6.6 to Sonoma 14.5 and now it works! The docstrings for `queryClient.py` also look good here, so I'll close PR #68 . Thanks @iglesu !

**Comment by rnikutta** [comment]
That's great!

@iglesu This is a separate issue, but to build a new version of the datalab CLI pip package, we should probably updated this workflow: https://github.com/astro-datalab/datalab/blob/master/.github/workflows/datalab_pip.yml

Can we add Python versions? At least 3.10, but maybe also 3.11?

And update the version (I think we agreed by tagging master (after merging), and then triggering a release?

Thanks!

File: requirements.txt
Comment (iglesu) [reply requirements.txt:1]: it's a tight rope to walk. I broaden the version range as far as it would work for the given python version, OS etc, but if you have a wider version recipe that works, please do share it and we'll test it.

**Comment by iglesu** [comment]
> That's great!
> 
> @iglesu This is a separate issue, but to build a new version of the datalab CLI pip package, we should probably updated this workflow: https://github.com/astro-datalab/datalab/blob/master/.github/workflows/datalab_pip.yml
> 
> Can we add Python versions? At least 3.10, but maybe also 3.11?
> 
> And update the version (I think we agreed by tagging master (after merging), and then triggering a release?
> 
> Thanks!

very good point, I think we are going to support 3.9 3.10 and 3.11 for now. I couldn't get 3.8 to install on a Mac M2, so I think I'm going to give up on it.

**Comment by rnikutta** [comment]
I think that's a good decision. On the new gp13 there won't be 3.8 either. (and I think there isn't one on the current gp12 even)

**Comment by iglesu** [comment]
@rnikutta I see gp12 still uses python3.8, I can add the python3.8 to the requirements file and do a hack to the setup.py to allow python 3.8 only in linux. This change will push this PR to next week. Unless we rather wait for gp13. Please advice.

**Comment by rnikutta** [comment]
@iglesu I suggest we wait for gp13, and don't bother with getting 3.8 to work, which would be needed for a few days only anyways. Let's hold merging of this PR for a few more days. gp13 is around the corner. Thanks!

**Pull request merged**
